kornia/enhance/jpeg.py

```diff
@@ -0,0 +1,573 @@
+        >>> jpeg_quality = torch.tensor((99.0, 25.0, 1.0), requires_grad=True)
+        >>> img_jpeg = jpeg_encode_decode_differentiable(img, jpeg_quality)
+        >>> img_jpeg.sum().backward()
+    """
+    # Check that inputs are tensors
+    KORNIA_CHECK_IS_TENSOR(image_rgb)
+    KORNIA_CHECK_IS_TENSOR(jpeg_quality)
+    # Get device and dtype
+    dtype: Dtype = image_rgb.dtype
+    device: Device = image_rgb.device
+    # Use default QT if QT is not given
+    quantization_table_y = _get_default_qt_y(device, dtype) if quantization_table_y is None else quantization_table_y
+    quantization_table_c = _get_default_qt_c(device, dtype) if quantization_table_c is None else quantization_table_c
+    KORNIA_CHECK_IS_TENSOR(quantization_table_y)
+    KORNIA_CHECK_IS_TENSOR(quantization_table_c)
+    # Check shape of inputs
+    KORNIA_CHECK_SHAPE(image_rgb, ["B", "3", "H", "W"])
+    KORNIA_CHECK(
+        (image_rgb.shape[2] % 16 == 0) and (image_rgb.shape[3] % 16 == 0),
+        f"image dimension must be divisible by 16. Got the shape {image_rgb.shape}.",
+    )
+    KORNIA_CHECK_SHAPE(jpeg_quality, ["B"])
+    # Add batch dimension to quantization tables if needed
+    if quantization_table_y.ndim == 2:
+        quantization_table_y = quantization_table_y.unsqueeze(dim=0)
+    if quantization_table_c.ndim == 2:
+        quantization_table_c = quantization_table_c.unsqueeze(dim=0)
+    # Check resulting shape of quantization tables
+    KORNIA_CHECK_SHAPE(quantization_table_y, ["B", "8", "8"])
+    KORNIA_CHECK_SHAPE(quantization_table_c, ["B", "8", "8"])
+    # Check value range of JPEG quality
+    KORNIA_CHECK(
+        (jpeg_quality.amin().item() >= 0.0) and (jpeg_quality.amax().item() <= 99.0),
+        f"JPEG quality is out of range. Expected range is [0, 99], "
+        f"got [{jpeg_quality.amin().item()}, {jpeg_quality.amax().item()}]. Consider clipping jpeg_quality.",
+    )
+    # Get original shape
+    H, W = image_rgb.shape[2:]
+    # Quantization tables to same device and dtype as input image
+    quantization_table_y = quantization_table_y.to(device, dtype)
+    quantization_table_c = quantization_table_c.to(device, dtype)
+    # Perform encoding
+    y_encoded, cb_encoded, cr_encoded = _jpeg_encode(
+        image_rgb=image_rgb,
+        jpeg_quality=jpeg_quality,
+        quantization_table_c=quantization_table_c,
+        quantization_table_y=quantization_table_y,
+    )
+    image_rgb_jpeg: Tensor = _jpeg_decode(
+        input_y=y_encoded,
+        input_cb=cb_encoded,
+        input_cr=cr_encoded,
+        jpeg_quality=jpeg_quality,
+        H=H,
+        W=W,
+        quantization_table_c=quantization_table_c,
+        quantization_table_y=quantization_table_y,
+    )
+    # Clip coded image
+    image_rgb_jpeg = _differentiable_clipping(input=image_rgb_jpeg, min=0.0, max=255.0)
+    return image_rgb_jpeg
+
+
+class JPEGEncodeDecodeDifferentiable(Module):
```

Comment:
Not quite sure this is very valuable to provide since any pytorch user can easily create its own Module in any possible shape of the constructor. E.g if had to design the nn.Module of this function i'd put all the extra args (or maybe just a couple) into the module constructor so that the forward expect only the input image. But this is i guess of application specific, so i wouldn't support module (plus one less thing to maintain). Open to opinions @johnnv1 @shijianjian

Reply:
Good point! People might want to learn the quantization tables. So we could initialize the module with quantization tables. Let me know what you think about this. Maybe in a future PR?

Reply:
I added the option to specify the quantization tables (fixed and as parameters) in constructor, just in case someone wants to learn the quantization tables.